ACL2022/Models/are_models

```diff
@@ -0,0 +1,324 @@
+      input_segments.append(inputs['token_type_ids'])
+    return (np.asarray(input_ids, dtype='int32'), np.asarray(input_masks, dtype='int32'), np.asarray(input_segments, dtype='int32'))
+
+
+
+  def build(self, bias=0):
+        in_id = Input(shape=(self.max_seq_length,), name="input_ids", dtype='int32')
+        in_mask = Input(shape=(self.max_seq_length,), name="input_masks", dtype='int32')
+        in_segment = Input(shape=(self.max_seq_length,), name="segment_ids", dtype='int32')
+        bert_inputs = [in_id, in_mask, in_segment]
+
+        bert_output = self.BERT(bert_inputs).last_hidden_state
+        bert_output = bert_output[:,0,:] #take only the ['CLS'] token
+        pred = Dense(1, activation = self.dense_activation, name = 'classification')(bert_output)
+        self.model = tensorflow.keras.models.Model(inputs=bert_inputs, outputs=pred)
+        self.model.compile(loss=self.loss,
+                      optimizer=tensorflow.keras.optimizers.Adam(learning_rate=self.lr),
+                      metrics=self.METRICS)
+        if self.show_summary:
+            self.model.summary()
+  
+  
+  def fit(self, train_X, train_y, dev_X, dev_y, bert_weights=None, class_weights={0: 1, 1: 1}, pretrained_embeddings=None):
+        train_input = self.to_bert_input(train_X)
+        dev_input = self.to_bert_input(dev_X)
+        self.build()
+        if bert_weights is not None:
+            self.model.load_weights(bert_weights)
+        history = self.model.fit(train_input,
+                       train_y,
+                       validation_data=(dev_input, dev_y),
+                       epochs=self.epochs,
+                       callbacks=[self.earlystop],
+                       batch_size=self.batch_size,
+                       class_weight=None 
+                       )
+        return history
+
+  def predict(self, texts):
+        test_input = self.to_bert_input(texts)
+        predictions = self.model.predict(test_input)        
+        print('Stopped epoch: ', self.earlystop.stopped_epoch)
+        return predictions
+  
+  
+   #take the average (of all heads) of the last layer for the ['CLS'] to  
+  def get_attentions(self, test_texts, layer = 12, batch_size = 100):
+    atts = []
+    counter = 0 
+    for i in range(0, len(test_texts)//batch_size):
+      texts = test_texts[i*batch_size: (i+1)*batch_size]
+      input = self.to_bert_input(texts)
+      bert_atts = tensorflow.math.reduce_mean(self.BERT(input).attentions[layer-1][:,:], axis=1)[:,0,:]
+      atts.append(bert_atts)
+      counter = i
+    counter +=1
+    if  len(test_texts)%batch_size != 0:
+      texts = test_texts[counter*100:]
+      input = self.to_bert_input(texts)
+      bert_atts = tensorflow.math.reduce_mean(self.BERT(input).attentions[layer-1][:,:], axis=1)[:,0,:]
+      atts.append(bert_atts)
+    atts = [item for sublist in atts for item in sublist] #flatten
+    return atts
+
+  
+  def finetune_att_threshold(self, dev_texts, dev_position):
```

Comment:
same as above - pull this out to a single function?

Reply:
fixed